[[ ReturnValue ]] Provide new return forms to return values and errors

runrevmark opened the pull request:
This patch adds two new forms of the 'return' command:

```
return <value> for value
return <value> for error
```

These forms act slightly differently, depending on whether the handler
calling them is a command or a function handler.

If the handler performing the return is a command handler then:

The 'return for value' form, sets 'it' in the caller to <value> and
sets 'the result' to empty.

The 'return for error' form, sets 'it' in the caller to empty and
sets 'the result' to <value>.

If the handler performingthe return is a function handler then:

The 'return for value' form, returns <value> as the return value of
the function call and sets 'the result' to empty.

The 'return for error' form, returns empty as the return value of
the function call and sets 'the result' to <value>.

These forms of return are designed to be used by script library
functions to allow them to have the same ability as built-in engine
functions and commands - namely the ability to return a value (in it
for commands, or return value for functions) _or_ return a status
(in the result).

runrevmark committed:
[[ ReturnValue ]] Provide new return forms to return values and errors

This patch adds two new forms of the 'return' command:

    return value <value>
    return error <value>

These forms act slightly differently, depending on whether the handler
calling them is a command or a function handler.

If the handler performing the return is a command handler then:

The 'return value' form, sets 'it' in the caller to <value> and
sets 'the result' to empty.

The 'return error' form, sets 'it' in the caller to empty and
sets 'the result' to <value>.

If the handler performingthe return is a function handler then:

The 'return value' form, returns <value> as the return value of
the function call and sets 'the result' to empty.

The 'return error' form, returns empty as the return value of
the function call and sets 'the result' to <value>.

These forms of return are designed to be used by script library
functions to allow them to have the same ability as built-in engine
functions and commands - namely the ability to return a value (in it
for commands, or return value for functions) *or* return a status
(in the result).

montegoulding commented:
I can only think of one command `read` that sets both to a value and it is a bit of an anomaly.

runrevmark commented:
@montegoulding: Do you have an example of when it does that as I'm not sure it should? I know the 'answer file with type' command also sets both.

montegoulding commented:
It sets the result to `eof` when it hits it regardless of whether the script tries to read past it. I guess we can add `dispatch` to the list also.

livecodeali commented on docs/dictionary/control_st/return.lcdoc:
```diff
@@ -2,39 +2,142 @@ Name: return
+       return value 1
+    end simpleCommand
+    on testSimpleCommand
+       simpleCommand
```
This is probably causing an issue, you might have to change to `get simpleCommand()`. Unless `simpleCommand` is meant to be a command.

montegoulding commented:
I'd say it's the variation in `[ ]` rather than `{ }`

montegoulding commented:
It should be in both as both variation and optional

montegoulding commented on docs/dictionary/control_st/return.lcdoc:
```diff
@@ -2,39 +2,142 @@ Name: return
 Type: control structure
 
-Syntax: return <value> 
+Syntax: return [ value | error ] <value>
```
Change to `Syntax: return [{ value | error }] <value>` should pass the docs test

montegoulding replied:
Hmm... it says it passes the docs test and fails the IDE tests, however, I think the above should fail the docs test so I will investigate why it's not.

montegoulding replied:
Aha!... my syntax checker is broken because:

```
not "return [ value | error ] <value>" contains "port" is false 
```

while

```
not ("return [ value | error ] <value>" contains "port") is true
```

A mistake I should not have made after so long scripting in LC!

montegoulding replied:
BTW I can't help wondering if the socket ID should be a single parameter in syntax declarations with detail of `IP:port|ID` in the param description to avoid the special case `| ID` having to be dealt with by anything parsing syntax declarations. I'm happy to make this change to the relevant docs and remove the condition from the syntax checker if people agree. There's only 4 docs that separate `socketID` into the components as separate parameters and it could give users the mistaken impression that you can declare a socket ID as `"127.0.0.1":"80"|"myserver"`.

runrevmark commented:
Okay - the PR no longer crashes by itself. It is affected by the ObjectHandle problem which is fixed in another PR awaiting merge - the latter causes some IDE tests to fail somewhat non-deterministically.

There is some more work needed here:

```
   return value <expr>
```

is ambiguous with:

```
  return value(<expr>)
```

Also, 'dispatch' needs some thought as it currently trounces 'it' making it impossible to dispatch commands which set it.

runrevmark commented:
Current thoughts on how to resolve the ambiguity is to use:

```
    return <expr> for value
    return <expr> for error
```

This is unambigous as it augments the existing `return` syntax rather than changing it.

runrevmark commented:
In regards to 'dispatch' - then we need a spike to work out how to sort that out. At the moment you can't dispatch to a command which uses the execContexts trick to set it as dispatch clobbers it, so there isn't a loss in functionality with the new 'return' forms.

runrevmark committed:
[[ ReturnValue ]] Change syntax to avoid ambiguity

This patch changes the syntax of the new return forms so they
don't conflict with functions called 'value' or 'error'.

The syntax is now:

    return <expr> for error
    return <expr> for value

livecodefraser commented:
@livecode-vulcan review okay e10ba02

livecode-vulcan commented:
💙 review by @livecodefraser ok e10ba028cabd248811dd7eeb0f30c4e37c367b29

livecode-vulcan commented:
😎 test success e10ba028cabd248811dd7eeb0f30c4e37c367b29
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2672)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2682)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2667)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2683)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2689)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2695)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2707)

mwieder commented:
Coming back into this late (I haven't been paying enough attention lately)... what about the syntax
`return <expr> as value
`
`
    return <expr> as error
`
rather than (or in addition to, although I know you all are allergic to synonyms)
`return <expr> for value
`
`
    return <expr> for error
`
as I think that makes it a more natural-language form.

runrevmark commented:
@mwieder: Apologies, I missed that comment - it probably is wise to avoid commenting on closed pull requests and raise suggestions elsewhere after they've been closed. Once a PR has been merged, it isn't entirely obvious when further comments appear on it as focus moves to the currently opened PRs!

Anyway, the syntax you proposed was considered - however, the form:

```
   X as Y
```

Has long been reserved (in my head at least) as a type conversion operator; so using it in this context now would mean it could not be used for that in the future.

mwieder commented:
(Commenting yet again in a closed PR) Thanks for the response. Yes, I was rather ticked off at this getting ignored. I commented after the PR was merged but before the release, so I thought maybe I still had the chance to change things before this got unleashed on the world. At any rate, your explanation makes sense, and possibly even opens the door for something like
`return tVar as double`, so even with all the consternation this is causing on the lists, I'm reasonably happy with Things As They Are.

runrevmark commented:
@mwieder: (Also commenting on a closed PR) We should perhaps lock conversation when the PR is closed, because unless there is an explicit @someone in the comment, it just gets lost in the large sea of notifications which GitHub sends out (and even with an @ comment they can still get lost). In regards to 'as' - then yes - it is reserved for a future facility for being able to explicitly change the 'type' of a value. At the LiveCode Script level I'd perhaps see this as a much nicer replacement for things like:

```
    put tString + 0 into tNumberFromString -- tString as number
    put tNumber & empty into tStringFromNumber -- tNumber as string
```

i.e. Ensure that a given value is a specific dynamic type rather than the type it happens to be.
